devito/interfaces.py

```diff
@@ -1,81 +1,257 @@
+        _SymbolCache[cls] = obj
+
+    @classmethod
+    def indices(cls, shape):
+        """Abstract class method to determine the default dimension indices."""
+        raise NotImplementedError('Abstract class `SymbolicData` does not have default indices')
+
+
+class DenseData(SymbolicData):
+    """Data object for spatially varying data that acts as a Function symbol
+
+    :param name: Name of the resulting :class sympy.Function: symbol
+    :param shape: Shape of the spatial data grid
+    :param dtype: Data type of the buffered data
+
+    Note: :class DenseData: objects are assumed to be constant in time and
+    therefore do not support time derivatives. Use :class TimeData: for
+    time-varying griad data.
+    """
+    def __init__(self, *args, **kwargs):
+        if self._cached():
+            # Initialise instance from symbol cache
+            SymbolicData.__init__(self)
+            return
+        else:
+            self.name = kwargs.get('name')
+            self.shape = kwargs.get('shape')
+            self.dtype = kwargs.get('dtype', np.float32)
+            self._data = kwargs.get('_data', None)
+            self.initializer = None
+            # Store new instance in symbol cache
+            self._cache_put(self)
+
+    @classmethod
+    def indices(cls, shape):
+        """Return the default dimension indices for a given data shape
+
+        :param shape: Shape of the spatial data
+        """
+        _indices = [x, y, z]
+        return _indices[:len(shape)]
+
+    @property
+    def indexed(self):
+        """Returns base symbol as sympy.IndexedBase"""
+        return IndexedBase(self.name, shape=self.shape)
+
+    def indexify(self):
+        """Convert base symbol and dimensions to indexed data accesses"""
+        indices = [a.subs({h: 1, s: 1}) for a in self.args]
+        return self.indexed[indices]
 
     def set_initializer(self, lambda_initializer):
         assert(callable(lambda_initializer))
         self.initializer = lambda_initializer
 
     def _allocate_memory(self):
-        self.pointer = aligned(np.zeros(self.shape, self.dtype, order='C'), alignment=64)
+        self._data = aligned(np.zeros(self.shape, self.dtype, order='C'), alignment=64)
 
     @property
     def data(self):
-        if self.pointer is None:
+        if self._data is None:
             self._allocate_memory()
-        return self.pointer
+        return self._data
 
     def initialize(self):
         if self.initializer is not None:
             self.initializer(self.data)
         # Ignore if no initializer exists - assume no initialisation necessary
 
+    @property
+    def dx2(self):
+        """Symbol for the second derivative wrt the x dimension"""
+        return as_finite_diff(self.diff(x, x), [x - h, x, x + h])
+
+    @property
+    def dy2(self):
+        """Symbol for the second derivative wrt the y dimension"""
+        return as_finite_diff(self.diff(y, y), [y - h, y, y + h])
+
```

Comment:
This also needs to be parametrised on the order since the FD formulation depends on this returning a different result for different spatial and time orders.